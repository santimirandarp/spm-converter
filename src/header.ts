import type { BinaryData } from 'cheminfo-types';
import { IOBuffer } from 'iobuffer';

import { StringObject } from './types';
import { getKeyValues } from './utility/headerKeyVals';
import { getLength } from './utility/headerLength';
import { imageItem } from './utility/imageItem';

/**
 *
 * @param spmFile - spm file as a binary data
 * @returns - json object with the header information
 */
export function readHeader(spmFile: BinaryData): StringObject {
  const io = new IOBuffer(spmFile);
  const header: StringObject = { 'Ciao image list': [], 'Scanner list': [] };
  const regex = {
    keyValue: /.(?<key>.*?):\s*(?<value>.*)/,
    object: /^\\\*/,
  };

  /* seems safe amount to get to the data length */
  const length = getLength(io.readUtf8(150));
  io.offset = 0;

  const lines = io.readUtf8(length).split('\r\n');
  for (let i = 0; i < lines.length - 2; i++) {
    const currentLine = lines[i];
    const [, topLevelKey] = currentLine.split(regex.object);
    if (topLevelKey) {
      let [subObject, lastIRead] = getKeyValues(lines, ++i);
      if (topLevelKey === 'Ciao image list') {
        subObject = imageItem(subObject);
        header[topLevelKey].push(subObject);
      } else if (topLevelKey === 'Scanner list') {
        header[topLevelKey].push(subObject);
      } else {
        header[topLevelKey] = subObject;
      }
      i = lastIRead;
    }
  }
  return header;
}
