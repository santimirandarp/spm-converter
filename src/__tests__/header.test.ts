import { readFileSync } from 'fs';
import { join } from 'path';

import { IOBuffer } from 'iobuffer';

import { readHeader } from '../header';

const filePath = join(__dirname, 'data', 'AFM', 'AFM.spm');
const buffer = readFileSync(filePath);

describe('read header', () => {
  it('return the header', () => {
    const header = readHeader(buffer);
    console.log(header)
    expect(header['Ciao image list']).toHaveLength(2);
    expect(header['Scanner list']).toHaveLength(1);
    expect(header).toMatchSnapshot();
  });
});
