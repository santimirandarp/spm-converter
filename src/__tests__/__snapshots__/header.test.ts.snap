// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`read header return the header 1`] = `
{
  "Ciao image list": [
    {
      "Bytes/pixel": 4,
      "Data length": 1048576,
      "Data offset": 80960,
      "Data type": "AFM",
      "Start context": "OL",
    },
    {
      "Bytes/pixel": 2,
      "Data length": 1048576,
      "Data offset": 1129536,
      "Data type": "AFM",
      "Start context": "OL",
    },
  ],
  "Ciao scan list": {
    "AFM Current Sensitivity": "1 nA/V",
    "Adhesion Algorithm": "Threshold Crossing",
    "Adhesion Fit %": "0.1",
    "Analog2 High Voltage": "0",
    "Analog22 High Voltage": "0",
    "Aspect Ratio": "1:1",
    "Auto Sync Distance": "On",
    "Auto Sync Distance QNM": "On",
    "Auto X Sep": "0",
    "Auto Y Sep": "0",
    "Auto number": "2",
    "Auto pattern": "Linear",
    "Bias limit": "12 V",
    "Calculation Period": "60",
    "Calibrate Scan Line Shift": "Off",
    "Cantilever Angle": "8",
    "Capture direction": "Down",
    "Capture mode ": "Single",
    "Capture prelines": "50",
    "Captured Automatically": "0",
    "Clear Buffer for new Scan Area": "No",
    "Color table": "2",
    "Current range": "200 nA",
    "Cursor A": "0",
    "Cursor B": "200",
    "Cutoff Modulus": "0.3",
    "Cycles": "10",
    "Cycles to Average": "32",
    "Dark Lift Delay": "300",
    "Deformation Fit Region": "0.98",
    "Display level": "127",
    "Drive feedback": "Disabled",
    "Drive gain": "80",
    "Drive setpoint": "1",
    "Drive time": "0",
    "Dynamic Sync Distance": "Off",
    "Enable Optical Image Dewarping": "Off",
    "Engage Setpoint": "0.85",
    "Engage X offset": "0 nm",
    "Engage Y offset": "0 nm",
    "Engage Z avg.": "10",
    "Engage Z delay": "0.5",
    "Engage Z sens.": "12.5",
    "Engage step size": "0.138889",
    "Extended Current sens": "1 pA/V",
    "Feature Scan rate": "1",
    "Feature scan angle": "0",
    "Feature setpoint": "0.8",
    "Feature step size": "0.138889",
    "Feature threshold": "2000",
    "Fluid Cell": "No",
    "Force Data Points": "64",
    "Force Filter": "Off",
    "Freq. Control": "User-defined",
    "Freq. Control Inter": "User-defined",
    "Frequency Feedback": "Off",
    "Frequency Feedback Inter": "Off",
    "Gain Factor": "1.54",
    "Gain Normalization": "By DataBase",
    "Generic amp sens": "1 V/A",
    "Harmonics To Correct": "1",
    "Height engage": "Enabled",
    "Include Adhesion Force": "Yes",
    "Initial Setpoint": "0",
    "Interaction Time": "0.24",
    "Invert Torsional": "Yes",
    "Invert Vertical": "Yes",
    "Lift Height": "61.0726",
    "Lift rate": "4",
    "Line average": "No",
    "Line offset factor": "0",
    "Lines": "512",
    "Live Cell Background Removal": "Off",
    "MIROView Scan Direction": "0",
    "Max Force Fit Boundary": "80",
    "Max Harmonics": "20",
    "Medium": "Air",
    "Meter Display Mode": "Vertical/Horiz",
    "Meter display ctrl.": "Hidden",
    "Min Force Fit Boundary": "20",
    "Min int. gain": "3",
    "Minimum scan rate": "0.1",
    "Modulus Fit Model": "Hertzian (Spherical)",
    "NanoTrack X Offset Actual": "0",
    "NanoTrack X Offset Set": "0",
    "NanoTrack Y Offset Actual": "0",
    "NanoTrack Y Offset Set": "0",
    "Operating mode": "Image",
    "OptoXY Do Linear": "Off",
    "Output 1 Data Type": "Off",
    "Output 2 Data Type": "Off",
    "PF Extended Current sens": "20 pA/V LBW",
    "Parameter Update FB": "On",
    "Parameter Update Retract": "Disabled",
    "Parameter select": "Main",
    "Pass 2 size": "10000 nm",
    "Pass 3 size": "5000 nm",
    "Pass 4 size": "2000 nm",
    "Pass delay": "5",
    "Peak Force Amplitude": "150",
    "Peak Force Engage Amplitude": "150",
    "Peak Force Engage Setpoint": "0.15",
    "Potential Feedback": "Off",
    "Potential Feedback Inter": "Off",
    "Potential Type": "Off",
    "Potential Type Inter": "Off",
    "QNM Calibration": "DefaultByZ",
    "QNM scan line shift": "0",
    "Rectilinear Scan": "Disabled",
    "Rotate Ang.": "0",
    "SECPM sens": "100 mV/V",
    "Sample Points": "64",
    "Sample Poisson's Ratio": "0.3",
    "Sample Rate": "10",
    "Samps/line": "512",
    "Scan Acceptable Threshold": "0.2",
    "Scan Max Optimizing Time": "30",
    "Scan Rate": "0.994238",
    "Scan Size": "5000 nm",
    "Scan Target Threshold": "0.005",
    "Scan line shift": "0",
    "ScanAsyst Auto Control": "On",
    "ScanAsyst Auto Gain": "On",
    "ScanAsyst Auto Scan Rate": "On",
    "ScanAsyst Auto Setpoint": "On",
    "ScanAsyst Auto Z Limit": "On",
    "ScanAsyst AutoConfig Frames": "0",
    "ScanAsyst Noise Threshold": "0.8",
    "ScanAsyst Setup": "Allow",
    "Scope dualtrace": "Dual",
    "Setpoint Units": "nm",
    "Show Frame Rate Control": "Hidden",
    "Show HSDC Button": "Both",
    "Show Point & Shoot Button": "Both",
    "Show Potential Controls": "Neither",
    "Show Repulsion Time Indicator": "Hidden",
    "Show Thermal Tune Button": "Both",
    "Signal/Noise Threshold": "2",
    "Skip Offset Optimization": "No",
    "Stage Optics": "-1175.38",
    "Stage X": "81663",
    "Stage Y": "147744",
    "Stage Z": "-18594.2",
    "Start size": "13199.8 nm",
    "Step XY period": "0.005",
    "Step XY size": "300",
    "Step size": "3",
    "Stop motor": "Disabled",
    "Strip Chart Rate": "500",
    "Strip Chart Size": "100",
    "Strip chart ": "Disabled",
    "Sync Distance": "25.6437",
    "Sync Distance New": "25.6437",
    "Sync Distance QNM": "25.6437",
    "Tip Frame Count": "14",
    "Tip Half Angle": "0",
    "Tip Offset X": "0",
    "Tip Offset Y": "0",
    "Tip Radius": "5",
    "Tip Scanned Distance": "143403",
    "Tip Serial Number": "Unknown",
    "Tip Velocity": "11.0471",
    "Top Fit Region": "0.3",
    "Torsional Freq": "950",
    "Torsional Q": "1000",
    "Units": "Metric",
    "Unload Fit Region": "0.7",
    "X Offset": "0 nm",
    "X Position": "0",
    "X drift": "0",
    "X round": "0.1",
    "XY Closed Loop": "Off",
    "XY Drift Correction": "Enabled",
    "XY Feedback Control": "Digital",
    "XY Move Tip Pos": "Surface",
    "Y Offset": "0 nm",
    "Y Position": "0",
    "Y disable": "Enabled",
    "Y drift": "0",
    "Z Closed Loop": "Off",
    "Z Range": "2.53271",
    "Z Sensor Correction": "Enabled",
    "Z offset": "0",
    "Z tolerance": "1180.74",
    "Zoom in at Constant": "Scan Rate",
    "bidirectional scan": "Disabled",
  },
  "Equipment list": {
    "Description": "Dimension Fastscan",
    "Scanner file": "1bd0e_fsz133.scn",
    "Vision": "uEye UI148xLE-C #1",
  },
  "Fast Scan list": {
    "All Pass Filter": "1",
    "Check for active tip": "Enabled",
    "Contact Engage Fast I Gain": "5",
    "Contact Engage Fast P Gain": "5",
    "Contact Gain": "123",
    "Contact Maximum Scan Rate": "4",
    "Contact Minimum Scanning I Gain": "0.1",
    "Contact Minimum Scanning P Gain": "0.1",
    "Determine Contact Gain": "Enabled",
    "Fast Integral Gain": "20",
    "Fast Interleave Igain": "0",
    "Fast Interleave Pgain": "0",
    "Fast Proportional Gain": "20",
    "High Pass Filter": "1",
    "Interleave Mode": "Disabled",
    "Lift Mode": "Disabled",
    "Low Pass Filter": "1",
    "Phase Shift": "508",
    "Q-Control Gain": "0",
    "QC Interleave Gain": "0",
    "QC Interleave Mode": "Disabled",
    "QC Lift Gain": "0",
    "QC Lift Mode": "Disabled",
    "Tapping Engage Fast I Gain": "10",
    "Tapping Engage Fast P Gain": "10",
    "Tapping Maximum Scan Rate": "3",
    "Tapping Minimum Scanning I Gain": "0.1",
    "Tapping Minimum Scanning P Gain": "0.1",
    "Z Sensor Preamp Gain": "2.52324",
    "Z Sensor Preamp Offset": "-0.0457875",
  },
  "File list": {
    "Data length": "80960",
    "Date": "04:49:10 PM Wed Dec 15 2021",
    "Start context": "OL2BIG",
    "User Rating": "0",
    "Version": "0x09400202",
  },
  "Scanner list": [
    {
      "@Sens. CurrentSens": "V -10.00000 nA/V",
      "@Sens. SECPMSens": "V 100.0000 mV/V",
      "@Sens. Xsensor": "V 3490.643 nm/V",
      "@Sens. Ysensor": "V 3422.620 nm/V",
      "@Sens. Zsens": "V 4.380000 nm/V",
      "Adaptive Max scan size": "34",
      "Adaptive Minimal Scan Rate": "0.2",
      "Adaptive Minimal Scan Size": "10",
      "Adaptive ScanSizeExpRatio": "0.5",
      "Adaptive X DC Gain": "0.6",
      "Adaptive XY Feedback Scanner": "Yes",
      "Adaptive Y DC Gain": "0.6",
      "Adaptive x CutOffFreq": "400",
      "Adaptive x Feedback BW": "100",
      "Adaptive x PlantGain": "2",
      "Adaptive x RefPolarity": "-1",
      "Adaptive x SensorNoise": "320",
      "Adaptive x TransferFuncFileName": "040810_morpheus_head_a1_noz_a1-fx.tfn",
      "Adaptive y CutOffFreq": "400",
      "Adaptive y Feedback BW": "100",
      "Adaptive y PlantGain": "2",
      "Adaptive y RefPolarity": "-1",
      "Adaptive y SensorNoise": "320",
      "Allow rotation": "Allow",
      "ClosedLoop x igain": "60",
      "ClosedLoop y igain": "60",
      "Extended offset der": "-12",
      "Fast arg derate": "0",
      "Fast cal freq": "2.50401",
      "File name": "1bd0e_fsz133.scn",
      "Full Range Sensor Gain": "1.4",
      "Has I2C": "Yes",
      "Has Tip Bias": "No",
      "Head Type": "Morpheus",
      "Illumination Range": "1023",
      "Illumination Scale Factor": "0.5",
      "Max scan size": "34.4",
      "Max tip velocity": "4500",
      "OLXY Scanner": "Yes",
      "OptoXY orthocal_x": "0",
      "OptoXY orthocal_xx": "0",
      "OptoXY orthocal_xy": "0",
      "OptoXY orthocal_y": "0",
      "OptoXY orthocal_yy": "0",
      "OptoXY orthog fitted": "-0.0383972",
      "OptoXY orthogonality": "-0.0383972",
      "OptoXY pgain/igain": "0.5",
      "OptoXY scan angle cor": "0.00174533",
      "OptoXY x sens fitted": "195",
      "OptoXY x sensitivity": "195",
      "OptoXY xsenscal_x": "0",
      "OptoXY xsenscal_xx": "0",
      "OptoXY xsenscal_xxx": "0",
      "OptoXY xsenscal_xxxx": "0",
      "OptoXY xsenscal_xxxy": "0",
      "OptoXY xsenscal_xxy": "0",
      "OptoXY xsenscal_xxyy": "0",
      "OptoXY xsenscal_xy": "0",
      "OptoXY xsenscal_xyy": "0",
      "OptoXY xsenscal_xyyy": "0",
      "OptoXY xsenscal_y": "0",
      "OptoXY xsenscal_yy": "0",
      "OptoXY xsenscal_yyy": "0",
      "OptoXY xsenscal_yyyy": "0",
      "OptoXY y sens fitted": "191.2",
      "OptoXY y sensitivity": "191.2",
      "OptoXY ysenscal_x": "0",
      "OptoXY ysenscal_xx": "0",
      "OptoXY ysenscal_xxx": "0",
      "OptoXY ysenscal_xxxx": "0",
      "OptoXY ysenscal_xxxy": "0",
      "OptoXY ysenscal_xxy": "0",
      "OptoXY ysenscal_xxyy": "0",
      "OptoXY ysenscal_xy": "0",
      "OptoXY ysenscal_xyy": "0",
      "OptoXY ysenscal_xyyy": "0",
      "OptoXY ysenscal_y": "0",
      "OptoXY ysenscal_yy": "0",
      "OptoXY ysenscal_yyy": "0",
      "OptoXY ysenscal_yyyy": "0",
      "Optoxy FitVersion": "V1",
      "Orthogonality": "0",
      "Part Number": "931-014-972",
      "Piezo cal": "440",
      "Piezo size": "G",
      "Retracted offset der": "14",
      "Rounding Minimum": "0",
      "Scanner type": "Dim 4000",
      "Serial Number": "1bd0e",
      "Slow arg derate": "0.005",
      "Slow cal freq": "4.89064",
      "StrainZ Scanner": "Yes",
      "X arg": "3.4",
      "X center offset": "-10",
      "X derate": "0.07",
      "X mag": "1.2",
      "X mag1": "0.85",
      "X max": "440",
      "X offset sens": "85.25",
      "X sensitivity": "54.45",
      "X slow derate": "0.074",
      "X slow sensitivity": "54.44",
      "X slow-fast coupling": "0.12171",
      "X slow-fast coupling derating": "-0",
      "XY digital CL cutoff frequency": "800",
      "XY digital CL pgain/igain": "0.5",
      "XY digital CL xIgain": "40",
      "XY digital CL yIgain": "40",
      "Xs-Yf coupling": "0.00170358",
      "Xs-Yf coupling derating": "-1.23204e-005",
      "Y arg": "3.7",
      "Y center offset": "-10",
      "Y derate": "0.07",
      "Y fast derate": "0.074",
      "Y fast sensitivity": "54.44",
      "Y mag": "0.9",
      "Y mag1": "1.18",
      "Y max": "440",
      "Y offset sens": "85.25",
      "Y sensitivity": "54.45",
      "Y slow-fast coupling": "0.147242",
      "Y slow-fast coupling derating": "-0",
      "Ys-Xf coupling": "-0.0144994",
      "Ys-Xf coupling derating": "-3.29532e-005",
      "Z Part Number": "860-010-800",
      "Z engage position": "0",
      "Z max": "220",
      "Z min": "-220",
      "Z range sens": "6.8174",
      "Z sensor LinC1": "1.00984",
      "Z sensor LinC2": "0.0337723",
      "Z sensor LinC3": "0.0129232",
      "Z sensor LinC4": "0.0242272",
      "Z sensor LinOC": "0.0252458",
      "Z sensor Linearization": "Yes",
      "Z sensor gain factor": "1.6",
      "Z sensor offset gain": "3.36",
      "Z sensor sens": "426.15",
      "Z serial": "fsz133",
      "Z to X coupling": "-10",
      "Z to X coupling FF": "-9",
      "Z to Y coupling": "-0.5",
      "Z to Y coupling FF": "-1",
    },
  ],
}
`;
